src/escapei.rs

```diff
@@ -212,8 +212,66 @@ where
     }
 }
 
+/// Returns the attribute value normalized as per the XML specification, using a custom
+/// entity resolver.
+///
+/// https://www.w3.org/TR/xml/#AVNormalize
+///
+/// Do not use this method with HTML attributes.
+///
+/// Escape sequences such as `&gt;` are replaced with their unescaped equivalents such as `>`
+/// and the characters `\t`, `\r`, `\n` are replaced with whitespace characters. A function
+/// for resolving entities can be provided as `resolve_entity`. Builtin entities will still
+/// take precedence.
+///
+/// This will allocate unless the raw attribute value does not require normalization.
+pub(crate) fn normalize_attribute_value_with<'a, 'entity>(
+    value: &'a str,
+    resolve_entity: impl Fn(&str) -> Option<&'entity str>,
+) -> Result<Cow<'a, str>, EscapeError> {
+    // TODO: avoid allocation when not needed
+    let mut normalized = String::with_capacity(value.len());
+
+    let attr = value.as_bytes();
+    let mut attr_iter = attr.iter().enumerate();
+
+    while let Some((idx, ch)) = attr_iter.next() {
+        match ch {
+            b' ' | b'\n' | b'\r' | b'\t' => normalized.push(' '),
+            b'&' => {
+                let end = idx
+                    + 1
+                    + attr_iter
+                        .position(|(_, c)| *c == b';')
+                        .ok_or_else(|| EscapeError::UnterminatedEntity(idx..attr.len()))?;
+                let entity = &attr[idx + 1..end]; // starts after the &
+                let entity_str = std::str::from_utf8(entity).expect("failed UTF-8 check");
+
+                if entity.starts_with(b"#") {
+                    let entity = &entity_str[1..]; // starts after the #
+                    let codepoint = parse_number(entity, idx..end)?;
+                    normalized.push_str(codepoint.encode_utf8(&mut [0u8; 4]));
+                } else if let Some(s) = named_entity(entity_str) {
+                    normalized.push_str(s);
+                } else if let Some(value) = resolve_entity(entity_str) {
+                    // TODO: recursively apply entity substitution
+                    normalized.push_str(&value);
+                } else {
+                    return Err(EscapeError::UnrecognizedSymbol(
+                        idx + 1..end,
+                        String::from_utf8(entity.to_vec()).expect("failed UTF-8 check"),
+                    ));
+                }
+            }
+            _ => normalized.push(*ch as char),
```

Comment:
This is incorrect -- all multibyte characters (for example, Russian letters) would be transformed into several ASCII characters. I've implemented correct algorithm in my https://github.com/Mingun/quick-xml/tree/attr-val-normalization branch, based on your work. It is not finished yet, because it does not process built-in entities. I've implemented recursion and do not allocate if there are nothing to replace.